align/pdk/finfet/transistor_array.py

```diff
@@ -13,30 +13,46 @@
 class MOSGenerator(CanvasPDK):
 
     def __init__(self, *args, **kwargs):
+        self.primitive_constraints = kwargs.get('primitive_constraints', [])
+
         super().__init__()
 
-        self.NEW_PARTIAL_ROUTING_FEATURE = os.getenv('PARTIAL_ROUTING', None) is not None
-        if self.NEW_PARTIAL_ROUTING_FEATURE:
+        self.style = None
+        self.PARTIAL_ROUTING = False
+        for const in self.primitive_constraints:
+            if const.constraint == 'generator':
+                if const.parameters is not None:
+                    self.style = const.parameters.get('style')
+                    self.PARTIAL_ROUTING = const.parameters.get('PARTIAL_ROUTING', False)
+
+        if os.getenv('PARTIAL_ROUTING', None) is not None:
+            self.PARTIAL_ROUTING = True
+
+        if self.PARTIAL_ROUTING:
             if not hasattr(self, 'metadata'):
                 self.metadata = dict()
             self.metadata['partially_routed_pins'] = {}
 
-    def addNMOSArray(self, x_cells, y_cells, pattern, vt_type, ports, **parameters):
-        self.mos_array_temporary_wrapper(x_cells, y_cells, pattern, vt_type, ports, **parameters)
-
-    def addPMOSArray(self, x_cells, y_cells, pattern, vt_type, ports, **parameters):
-        self.mos_array_temporary_wrapper(x_cells, y_cells, pattern, vt_type, ports, **parameters)
-
-    def mos_array_temporary_wrapper(self, x_cells, y_cells, pattern, vt_type, ports, **parameters):
-
         # Inject constraints for testing purposes
-        place_on_grid = os.getenv('PLACE_ON_GRID', False)
-        if place_on_grid:
+        place_on_grid = os.getenv('PLACE_ON_GRID', None)
+        if place_on_grid is not None:
             place_on_grid = json.loads(place_on_grid)
             if not hasattr(self, 'metadata'):
                 self.metadata = dict()
             self.metadata['constraints'] = place_on_grid['constraints']
 
+
+    def semantic(self):
```

Comment:
Is this still needed?

Reply:
Nope. Just removed it. But all PDKs not have to accept keyword arguments.